### 3.6 - Projeto Prático

In [14]:
import numpy as np
import pandas as pd
from Perceptron import Perceptron

# Para manter os dados consistentes, criei uma seed para sempre ter os mesmos pesos quando executado.
rng = np.random.default_rng(42)

In [15]:
# Importando a tabela do Apêndice I
# A tabela contém os dados de treinamento do Perceptron
dataframe = pd.read_csv('../data/tabela_apendice_i.csv', sep=',')
dataframe.head()

x1      x2       x3    d
0 -0.6508  0.1097   4.0009 -1.0
1 -1.4492  0.8896   4.4005 -1.0
2  2.0850  0.6876  12.0710 -1.0
3  0.2626  1.1476   7.7985  1.0
4  0.6418  1.0234   7.0427  1.0

In [16]:
# Função de ativação
def activation_func(u: float) -> int:
    return 1 if u >= 0 else -1

In [19]:
neurons = []
for _ in range(5):
    w = rng.random(4)
    neuron = Perceptron(w, activation_func)
    neuron.train(dataframe)
    neurons.append(neuron)

Peso inicial: [0.77395605 0.43887844 0.85859792 0.69736803]
Treinamento concluído na época: 425
Peso final: [-3.06604395  1.55241044  2.45939392 -0.73049397]
Peso inicial: [0.09417735 0.97562235 0.7611397  0.78606431]
Treinamento concluído na época: 437
Peso final: [-3.14582265  1.59203235  2.5394837  -0.74852969]
Peso inicial: [0.12811363 0.45038594 0.37079802 0.92676499]
Treinamento concluído na época: 439
Peso final: [-3.15188637  1.59407594  2.53302002 -0.74927901]
Peso inicial: [0.64386512 0.82276161 0.4434142  0.22723872]
Treinamento concluído na época: 395
Peso final: [-3.07613488  1.53326561  2.4679982  -0.73121128]
Peso inicial: [0.55458479 0.06381726 0.82763117 0.6316644 ]
Treinamento concluído na época: 418
Peso final: [-3.06541521  1.55375726  2.45920317 -0.7303736 ]


### Resultado dos treinamentos do Perceptron
| Treinamento | $w_{0}$ (inicial) | $w_{1}$ (inicial) | $w_{2}$ (inicial) | $w_{3}$ (inicial) | $w_{0}$ (final) | $w_{1}$ (final) | $w_{2}$ (final) | $w_{3}$ (final) | Épocas |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 1 | 0.7739 | 0.4388 | 0.8585 | 0.6973 | -3.0660 | 1.5524 | 2.4593 | -0.7304 | 425 |
| 2 | 0.0941 | 0.9756 | 0.7611 | 0.7860 | -3.1458 | 1.5920 | 2.5394 | -0.7485 | 437 |
| 3 | 0.1281 | 0.4503 | 0.3707 | 0.9267 | -3.1518 | 1.5940 | 2.5330 | -0.7492 | 439 |
| 4 | 0.6438 | 0.8227 | 0.4434 | 0.2272 | -3.0761 | 1.5332 | 2.4679 | -0.7312 | 395 |
| 5 | 0.5545 | 0.0638 | 0.8276 | 0.6316 | -3.0654 | 1.5537 | 2.4592 | -0.7303 | 418 |

In [38]:
samples = pd.read_csv("../data/amostras_oleo.csv", sep=',')

for i, row in samples.iterrows():
    inputs = np.array([-1, row.x1, row.x2, row.x3])

    predicts = []
    for neuron in neurons:
        predicts.append(neuron.predict(inputs))

    for j in range(5):
        samples.loc[i, f"y_T{j + 1}"] = predicts[j]

samples.head()
samples.to_csv("../data/amostras_oleo_classificacao.csv", sep=',')


### Amostras de óleo para validar o Perceptron
| Amostra | x1 | x2 | x3 | y_T1 | y_T2 | y_T3 | y_T4 | y_T5 |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 1 | -0.3665 | 0.0620 | 5.9891 | -1.0 | -1.0 | -1.0 | -1.0 | -1.0 |
| 2 | -0.7842 | 1.1267 | 5.5912 | 1.0 | 1.0 | 1.0 | 1.0 | 1.0 |
| 3 | 0.3012 | 0.5611 | 5.8234 | 1.0 | 1.0 | 1.0 | 1.0 | 1.0 |
| 4 | 0.7757 | 1.0648 | 8.0677 | 1.0 | 1.0 | 1.0 | 1.0 | 1.0 |
| 5 | 0.1570 | 0.8028 | 6.3040 | 1.0 | 1.0 | 1.0 | 1.0 | 1.0 |
| 6 | -0.7014 | 1.0316 | 3.6005 | 1.0 | 1.0 | 1.0 | 1.0 | 1.0 |
| 7 | 0.3748 | 0.1536 | 6.1537 | -1.0 | -1.0 | -1.0 | -1.0 | -1.0 |
| 8 | -0.6920 | 0.9404 | 4.4058 | 1.0 | 1.0 | 1.0 | 1.0 | 1.0 |
| 9 | -1.3970 | 0.7141 | 4.9263 | -1.0 | -1.0 | -1.0 | -1.0 | -1.0 |
| 10 | -1.8842 | -0.2805 | 1.2548 | -1.0 | -1.0 | -1.0 | -1.0 | -1.0 |